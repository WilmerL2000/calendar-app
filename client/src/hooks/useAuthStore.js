import { useDispatch, useSelector } from 'react-redux';
import { calendarApi } from '../api';
import {
  clearErrorMessage,
  onChecking,
  onLogin,
  onLogout,
} from '../store/auth/authSlice';
import { onLogoutCalendar } from '../store';

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  /**
   * This function attempts to log in a user by sending their email and password to an API, and if
   * successful, stores their token and user information in local storage.
   */
  const startLogin = async ({ email, password }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post('/auth', { email, password });

      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      console.log(error);
      dispatch(onLogout('Credenciales incorrectas'));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  /**
   * This function attempts to register a new user with an email, password, and name by making a POST
   * request to a calendar API and handling any errors that may occur.
   */
  const startRegister = async ({ email, password, name }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post('/auth/new', {
        email,
        password,
        name,
      });

      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      dispatch(onLogout(error?.response?.data?.msg));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  /**
   * This function checks if there is a valid authentication token stored in local storage and renews
   * it if possible, otherwise logs the user out.
   * @returns The function `checkAuthToken` is not returning anything explicitly, but it is likely
   * being used as an asynchronous action in a Redux application to check the validity of an
   * authentication token stored in the browser's local storage. Depending on the outcome of the token
   * check, the function may dispatch a `onLogin` or `onLogout` action to update the application state.
   */
  const checkAuthToken = async () => {
    const token = localStorage.getItem('token');
    if (!token) return dispatch(onLogout());

    try {
      const { data } = await calendarApi.get('/auth/renew');
      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      localStorage.clear();
      dispatch(onLogout());
    }
  };

  /**
   * The function clears local storage and dispatches two logout actions.
   */
  const startLogout = () => {
    localStorage.clear();
    dispatch(onLogout());
    dispatch(onLogoutCalendar());
  };

  return {
    //* Properties
    status,
    user,
    errorMessage,
    //* Methods
    startLogin,
    startRegister,
    checkAuthToken,
    startLogout,
  };
};
