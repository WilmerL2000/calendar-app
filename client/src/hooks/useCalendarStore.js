import { useDispatch, useSelector } from 'react-redux';
import {
  onAddNewEvent,
  onSetActiveEvent,
  onUpdateEvennt,
  onDeleteEvent,
} from '../store';

export const useCalendarStore = () => {
  const { events, activeEvent } = useSelector((state) => state.calendar);
  const dispatch = useDispatch();

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  const startSavingEvent = async (calendarEvent) => {
    // TODO: Backend

    if (calendarEvent._id) {
      dispatch(onUpdateEvennt({ ...calendarEvent }));
    } else {
      dispatch(onAddNewEvent({ ...calendarEvent, _id: new Date().getTime() }));
    }
  };

  const startDeletingEvent = () => {
    dispatch(onDeleteEvent());
  };

  return {
    //* Properties
    events,
    activeEvent,
    hasEventSelected: !!activeEvent,
    //* Methods
    setActiveEvent,
    startSavingEvent,
    startDeletingEvent,
  };
};
