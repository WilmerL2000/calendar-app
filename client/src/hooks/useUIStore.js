import { useDispatch, useSelector } from 'react-redux';
import {
  onCloseDateModal,
  onCloseDeleteModal,
  onOpenDateModal,
  onOpenDeleteModal,
  onSetActiveEvent,
} from '../store';

export const useUIStore = () => {
  const { isDateModalOpen, isDeleteModalOpen } = useSelector(
    (state) => state.ui
  );
  const dispatch = useDispatch();

  /**
   * The function `openDateModal` dispatches an action to open a date modal.
   */
  const openDateModal = () => {
    dispatch(onOpenDateModal());
  };

  /**
   * This function calls the onCloseDateModal action using dispatch.
   */
  const closeDateModal = () => {
    dispatch(onSetActiveEvent(null));
    dispatch(onCloseDateModal());
  };

  /**
   * This function calls the onOpenDeleteModal action using the dispatch method.
   */
  const openDeleteModal = () => {
    dispatch(onOpenDeleteModal());
  };

  /**
   * This function calls the onCloseDeleteModal action when the closeDeleteModal function is executed.
   */
  const closeDeleteModal = () => {
    dispatch(onCloseDeleteModal());
    dispatch(onSetActiveEvent(null));
  };

  const toggleDateModal = () => {
    isDateModalOpen
      ? dispatch(onOpenDateModal())
      : dispatch(onCloseDateModal());
  };

  return {
    //* Properties
    isDateModalOpen,
    isDeleteModalOpen,
    //* Methods
    openDateModal,
    closeDateModal,
    toggleDateModal,
    openDeleteModal,
    closeDeleteModal,
  };
};
