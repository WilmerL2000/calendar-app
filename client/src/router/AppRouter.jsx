import { Routes, Route, Navigate } from 'react-router-dom';
import { LoginPage } from '../auth';
import { CalendarPage, Loader } from '../calendar';
import { useAuthStore } from '../hooks';
import { useEffect } from 'react';

export const AppRouter = () => {
  const { status, checkAuthToken } = useAuthStore();

  useEffect(() => {
    checkAuthToken();
  }, []);

  if (status === 'checking') {
    return <Loader />;
  }

  return (
    <Routes>
      {status === 'authenticated' ? (
        <>
          <Route path="/" element={<CalendarPage />} />
          <Route path="/*" element={<Navigate to="/" />} />
        </>
      ) : (
        <>
          <Route path="/auth" element={<Navigate to="/auth/login" />} />
          <Route path="/auth/*" element={<LoginPage />} />
        </>
      )}
    </Routes>
  );
};
