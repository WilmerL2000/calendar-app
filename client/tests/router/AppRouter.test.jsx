import { useAuthStore } from '../../src/hooks/useAuthStore';
import { render, screen } from '@testing-library/react';
import { AppRouter } from '../../src/router/AppRouter';
import { MemoryRouter } from 'react-router-dom';

jest.mock('../../src/hooks/useAuthStore');

describe('Test on the AppRouter', () => {
  const mockCheckAuthToken = jest.fn();
  beforeEach(() => jest.clearAllMocks());

  test('It should show the loading screen and call check AuthToken ', () => {
    useAuthStore.mockReturnValue({
      status: 'checking',
      checkAuthToken: mockCheckAuthToken,
    });

    render(<AppRouter />);

    expect(screen.getByText('Loading...')).toBeTruthy();
    expect(mockCheckAuthToken).toHaveBeenCalled();
  });

  test('It must show the login if it is not authenticated ', async () => {
    useAuthStore.mockReturnValue({
      status: 'not-authenticated',
      checkAuthToken: mockCheckAuthToken,
    });

    const { container } = render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );

    expect(screen.getByText('Ingreso')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
