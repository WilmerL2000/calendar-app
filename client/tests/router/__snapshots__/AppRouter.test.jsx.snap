// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on the AppRouter It must show the login if it is not authenticated  1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="d-flex justify-content-center"
    >
      <div
        class="col-md-6 login-form-1 "
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Correo"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
          <div
            class="text-center mt-4"
          >
            <p>
              ¿No tienes una cuenta?
               
              <span
                class="text-primary cursor-pointer"
                role="button"
              >
                Registrate
              </span>
            </p>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
